rust/datafusion/src/execution/table_impl.rs

```diff
@@ -78,24 +108,210 @@ impl Table for TableImpl {
-                i
-            )));
+impl TableImpl {
+    /// Determine the data type for a given expression
+    fn get_data_type(&self, expr: &Expr) -> Result<DataType> {
+        match expr {
+            Expr::Column(i) => Ok(self.plan.schema().field(*i).data_type().clone()),
+            _ => Err(ExecutionError::General(format!(
+                "Could not determine data type for expr {:?}",
+                expr
+            ))),
         }
     }
-    Ok(Arc::new(Schema::new(fields)))
+
+    /// Create an expression to represent a named aggregate function
+    fn aggregate_expr(&self, name: &str, expr: &Expr) -> Result<Expr> {
+        let return_type = self.get_data_type(expr)?;
+        Ok(Expr::AggregateFunction {
+            name: name.to_string(),
+            args: vec![expr.clone()],
+            return_type,
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::execution::context::ExecutionContext;
+    use std::env;
+
+    #[test]
+    fn column_index() {
+        let t = test_table();
+        assert_eq!(0, t.column_index("c1").unwrap());
+        assert_eq!(1, t.column_index("c2").unwrap());
+        assert_eq!(12, t.column_index("c13").unwrap());
+    }
+
+    #[test]
+    fn select_columns() -> Result<()> {
+        let t = test_table();
+
+        let t2 = t.select_columns(vec!["c1", "c2", "c11"])?;
+
+        let plan = t2.to_logical_plan();
+
+        assert_eq!(
+            "Projection: #0, #1, #10\n  TableScan: aggregate_test_100 projection=None",
+            format!("{:?}", plan)
+        );
+
+        Ok(())
+    }
+
+    #[test]
+    fn select_expr() -> Result<()> {
+        let t = test_table();
+
+        let t2 = t.select(vec![t.col("c1")?, t.col("c2")?, t.col("c11")?])?;
+
+        let plan = t2.to_logical_plan();
+
+        assert_eq!(
+            "Projection: #0, #1, #10\n  TableScan: aggregate_test_100 projection=None",
+            format!("{:?}", plan)
+        );
+
+        Ok(())
+    }
+
+    #[test]
+    fn select_invalid_column() -> Result<()> {
+        let t = test_table();
+
+        match t.col("invalid_column_name") {
+            Ok(_) => panic!(),
+            Err(e) => assert_eq!(
+                "InvalidColumn(\"No column named \\\'invalid_column_name\\\'\")",
+                format!("{:?}", e)
+            ),
+        }
+
+        Ok(())
+    }
+
+    #[test]
+    fn aggregate() -> Result<()> {
+        let t = test_table();
+
+        let group_expr = vec![t.col("c1")?];
+        let c12 = t.col("c12")?;
+        let aggr_expr = vec![
+            t.min(&c12)?,
+            t.max(&c12)?,
+            t.avg(&c12)?,
+            t.sum(&c12)?,
+            t.count(&c12)?,
+        ];
+
+        let t2 = t.aggregate(group_expr, aggr_expr)?;
+
+        let plan = t2.to_logical_plan();
+
+        assert_eq!(
+            "Aggregate: groupBy=[[#0]], aggr=[[MIN(#11), MAX(#11), AVG(#11), SUM(#11), COUNT(#11)]]\n  TableScan: aggregate_test_100 projection=None",
```

Comment:
Is there a way to implement equality for query plan? It would be great if we can avoid equality check with string comparison.

Reply:
It is not possible (without code changes) to compare the plans. I have re-implemented these tests using a different strategy now, where each query build via the Table API is compared to the same query produced via SQL. This has made the tests much more concise and also detected a difference in the data type used for the LIMIT clause, which is resolved now.